compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64ArrayIndexOfOp.java

```diff
@@ -77,190 +90,301 @@ public AArch64ArrayIndexOfOp(int arrayBaseOffset, JavaKind valueKind, LIRGenerat
+        Register endOfString = asRegister(temp3);
+        Register refAddress = asRegister(temp4);
+        Register chunkToReadAddress = asRegister(temp5);
+        Register searchElementRegV = asRegister(vectorTemp1);
+        Register chunkPart1RegV = asRegister(vectorTemp2);
+        Register chunkPart2RegV = asRegister(vectorTemp3);
+        Register tmpRegV1 = asRegister(vectorTemp4);
+
+        Label matchInChunk = new Label();
+        Label searchByChunkLoopHead = new Label();
+        Label searchByChunkLoopTail = new Label();
+        Label processTail = new Label();
+        Label end = new Label();
+
+        int shiftSize = getShiftSize();
+        ElementSize eSize = ElementSize.fromSize(elementByteSize * Byte.SIZE);
+        /*
+         * Magic constant is used to detect byte offset of the starting position of the search
+         * element in the matching chunk.
+         */
+        long magicConstant = 0xc030_0c03_c030_0c03L;
+
+        /* 1. Duplicate the searchElement to 16-bytes */
+        masm.neon.dupVG(ASIMDSize.FullReg, eSize, searchElementRegV, searchElement);
         /*
-         * Set the current index (will be <= 0). Note that curIndex is not aligned to the chunk
-         * size.
+         * 2.1 Set endOfString pointing to byte next to the last valid element in the string and
+         * 'refAddress' pointing to the beginning of the last chunk.
          */
-        masm.sub(64, curIndex, fromIndex, endIndex);
-        if (isUTF16) {
-            masm.lsl(64, curIndex, curIndex, 1);
+        masm.add(64, endOfString, baseAddress, arrayLength, ShiftType.LSL, shiftSize);
+        masm.bic(64, refAddress, endOfString, 31L);
+        /* Set 'chunkToReadAddress' pointing to the chunk from where the search begins. */
+        masm.add(64, chunkToReadAddress, baseAddress, fromIndex, ShiftType.LSL, shiftSize);
+
+        masm.bind(searchByChunkLoopHead);
+        masm.cmp(64, refAddress, chunkToReadAddress);
+        masm.branchConditionally(ConditionFlag.LE, processTail);
+        masm.bind(searchByChunkLoopTail);
+        masm.sub(64, currOffset, chunkToReadAddress, baseAddress);
+
+        masm.fldp(128, chunkPart1RegV, chunkPart2RegV, AArch64Address.createImmediateAddress(128, AArch64Address.AddressingMode.IMMEDIATE_PAIR_POST_INDEXED, chunkToReadAddress, 32));
+        /* 3. Find searchElement in the 32-byte chunk */
+        masm.neon.cmeqVVV(ASIMDSize.FullReg, eSize, chunkPart1RegV, chunkPart1RegV, searchElementRegV);
+        masm.neon.cmeqVVV(ASIMDSize.FullReg, eSize, chunkPart2RegV, chunkPart2RegV, searchElementRegV);
+        /* Determining if there was a match. */
+        /* Combine two registers into 1 register */
+        masm.neon.orrVVV(ASIMDSize.FullReg, tmpRegV1, chunkPart1RegV, chunkPart2RegV);
+        /* Reduce 128-bit value to 32/64-bit value */
+        if (elementByteSize == 1) {
+            masm.neon.umaxvSV(ASIMDSize.FullReg, ElementSize.Word, tmpRegV1, tmpRegV1);
```

Comment:
On the neoverse timing charts I believe the Word variant is faster then the Byte variant